x86mmu: kernel: Validate existing APIs

akhileshupadhyay opened the pull request:
Testcase developed x86 specific to validate API for userbuffer.

Signed-off-by: Akhilesh Kumar Upadhyay <akhilesh.kumarx.upadhyay@intel.com>

tbursztyka commented on tests/kernel/userbufer_validate/src/main.c:
```diff
@@ -0,0 +1,45 @@
+/*test case nain entry*/
+void test_main(void *p1, void *p2, void *p3)
+{
+		ztest_test_suite(userbuffer_validate,
```
1 tab only for indentation on all these lines.

akhileshupadhyay replied:
1 tab indentation has been corrected.

tbursztyka commented on tests/kernel/userbufer_validate/src/userbuffer_validate.c:
```diff
@@ -0,0 +1,352 @@
+	/* user/supervisor test */
+	x86_mmu_set_pte_us_bit((void *)0x00405000, 0); /* supervisor */
+	status = _x86_mmu_buffer_validate((void *)0x00405000,
+			10, 0x1 | 0x2); /*  should fail */
```
parameters should be aligned to last '(' + 1 space, so here the 10, ... should be right belo (void *...

and the comment /* should fail */, put it on its own line.

Apply that rule everywhere

akhileshupadhyay replied:
breaking rule applied and corrected everywhere.

tbursztyka commented on tests/kernel/userbufer_validate/src/userbuffer_validate.c:
```diff
@@ -0,0 +1,352 @@
+{
+	zassert_true(buffer_rw_read() == TC_PASS, NULL);
+}
+void test_buffer_writeable_write(void)
```
always break a line before creating a new function. Apply that everywhere

akhileshupadhyay replied:
1 line break applied for all functions.

pfalcon reviewed (changes requested):
This patch is named as if it introduces "API to validate user buffer". But it does not, it just adds a test for an existing API. Please make sure that caption corresponds to the content.

Commit:
x86: mmu: kernel: Validate existing APIs

Testcase developed x86mmu specific, to validate
existing  APIs. This checks for the PDE/PTE set
on the address and returns if some violation occurs or not.

Signed-off-by: Akhilesh Kumar Upadhyay <akhilesh.kumarx.upadhyay@intel.com>

tbursztyka commented on tests/kernel/userbufer_validate/src/userbuffer_validate.c:
```diff
@@ -0,0 +1,380 @@
+	/* read write testing */
+	x86_mmu_set_pte_rw_bit((void *)0x00405000, 0);
+	status = _x86_mmu_buffer_validate((void *)0x00405000,
+			10, BUFF_WRITEABLE); /* should fail */
```
missed this line.

akhileshupadhyay replied:
the missed line corrected.

andyross commented:
Can we clean up the commit message?  This isn't x86 specific, it's actually PC-specific and assumes qemu's memory map.  The API being tested names itself something more like "x86 mmu" and not "user buffer".  Please explain exactly what functionality is being tested (it looks like it's simply setting PDE/PTE bits directly and then validating that they're still set?) in the commit message.

Also: wouldn't a more effective/portable/bitrot-resistant way to do this be to set the protection modes, and then try to access the pages in violation of that and check that the resulting exception is delivered correctly?

andrewboie commented on tests/kernel/userbufer_validate/src/userbuffer_validate.c:
```diff
@@ -0,0 +1,381 @@
+/* read write testing */
+static int multi_page_buffer_writeable(void)
+{
+	x86_mmu_set_pte_rw_bit((void *)0x00405000, 0);
```
why are these addresses hard-coded?
you need to programmatically select memory addresses to test.
there are linker variables defined to help you with this. see include/linker/linker-defs.h

akhileshupadhyay replied:
hard-coded addresses have been removed, also the linker variables have been used

tyrantlizard77 reviewed (changes requested):
You need to remove *all* the hard-coded memory addresses. This test shall not depend specifically on QEMU's memory map. Use defined linker variables to determine what memory ranges exist for various regions.

pfalcon reviewed (dismissed):
Thanks for updating the commit subject. (Didn't look further, just removing my -1.)

andrewboie commented on tests/kernel/userbufer_validate/src/userbuffer_validate.c:
```diff
@@ -0,0 +1,401 @@
+#include <linker/linker-defs.h>
+#include <ztest.h>
+
+void x86_mmu_set_pte_rw_bit(void *addr, bool value)
```
Why do we have these functions here?
We already have _x86_mmu_set_flags(). Use that instead.
You can make these functions call _x86_mmu_set_flags() if you want. But you should not be poking the page tables directly in this file.

andrewboie commented on tests/kernel/userbufer_validate/src/userbuffer_validate.c:
```diff
@@ -0,0 +1,401 @@
+	pte_addr->entry[pte].rw = (u32_t)value;
+}
+
+void x86_mmu_set_pte_us_bit(void *addr, bool value)
```
Don't need this function use _x86_mmu_set_flags()

andrewboie commented on tests/kernel/userbufer_validate/src/userbuffer_validate.c:
```diff
@@ -0,0 +1,401 @@
+	pte_addr->entry[pte].us = (u32_t)value;
+}
+
+void x86_mmu_set_pde_us_bit(void *addr, bool value)
```
Don't need this function use _x86_mmu_set_flags()

andrewboie commented on tests/kernel/userbufer_validate/src/userbuffer_validate.c:
```diff
@@ -0,0 +1,401 @@
+{
+	x86_mmu_set_pte_rw_bit((void *)__kernel_ram_start, 1);
+	status = _x86_mmu_buffer_validate((void *)__kernel_ram_start,
+					   10, 0x1);
```
Please do not use hard-coded constants when we have macros. Replace 0x1 with BUF_WRITEABLE here and elsewhere

akhileshupadhyay commented:
removed x86_mmu_set_pte_us_bit,  x86_mmu_set_pte_us_bit, x86_mmu_set_pde_us_bit  functions.
Used  _x86_mmu_set_flags() instead. Also the macro is used for masks and flags.

AdithyaBaglody reviewed (changes requested):
Test cases for Validating  multiple pages  are missing. Each of the  pages should be configured with different page table options(i.e RW/US).

tyrantlizard77 reviewed (changes requested):
This coverage is insufficient.

Right now, _x86_mmu_buffer_validate() is broken, it actually always returns true (nonzero) no matter what.

You are inconsistently checking if the return value is 0 or -EPERM. Please see PR #1441

tyrantlizard77 reviewed (changes requested):
I'm making some changes to this API so hold off on this PR until those land.

andrewboie commented:
hi, any update on this?
the APIs are now fixed.

akhileshupadhyay commented:
am working on it, update after fixing errors.

akhileshupadhyay commented:
@andrewboie. I have made the changes, included test coverage.

akhileshupadhyay commented:
@tbursztyka. The proper indentation in missed lines are restructured.

andrewboie commented:
@AdithyaBaglody can you review this?

andrewboie commented on tests/kernel/userbufer_validate/src/main.c:
```diff
@@ -0,0 +1,29 @@
+extern void test_buffer_user_rw_user(void);
+extern void test_buffer_user_rw_supervisor(void);
+/*test case nain entry*/
+void test_main(void *p1, void *p2, void *p3)
```
why is this in a separate file???

akhileshupadhyay replied:
This will make the code more legible and easy. Also some of the other codes follow same.

akhileshupadhyay commented:
@AdithyaBaglody . Changes related to multi pages. Removed all the hard coded values. Lines are aligned with one line gap after every new function starts.

akhileshupadhyay commented:
@AdithyaBaglody  I have added more test cases. This will cover multi pages test.

akhileshupadhyay commented:
@andrewboie could you please review the changes requested by you.

kumarvikash1 reviewed (approved):
Changes looks fine now, i can see "multi page buffer", "code coverage" and line breaking has been addressed.

kumarvikash1 commented:
@andrewboie  @tbursztyka   : Can you please review the changes. This PR is planned for R-1.10.

tbursztyka commented on tests/kernel/userbufer_validate/src/userbuffer_validate.c:
```diff
@@ -0,0 +1,524 @@
+static int buffer_rw_read(void)
+{
+	PRESET_PAGE_1_VALUE;
+	_x86_mmu_set_flags(ADDR_PAGE_1,
```
break a line after PRESET_... You do it sometime and sometime not. Apply that eveywhere

akhileshupadhyay replied:
@tbursztyka . Could you please check the style issue is addressed and fixed

tbursztyka reviewed (commented):
tiny style issue. rest is fine to me.
